File: apps/push-service/src/main/java/com/example/chat/push/infrastructure/sender/PushSender.java
package com.example.chat.push.infrastructure.sender;

/**
 * 푸시 발송 인터페이스
 *
 * 다형성 기반 설계 - 타입별로 구현체가 분리됩니다.
 */
public interface PushSender {
    boolean support(String pushType);

    void send(String targetUserId, String title, String content);
}

File: apps/push-service/src/main/kotlin/com/example/chat/push/infrastructure/sender/PushSender.kt
package com.example.chat.push.infrastructure.sender

interface PushSender {
    fun support(pushType: String): Boolean
    fun send(targetUserId: String, title: String, content: String)
}

import io.github.oshai.kotlinlogging.KotlinLogging
import org.springframework.stereotype.Component

private val logger = KotlinLogging.logger {}

@Component
class ToastPushSender : PushSender {
    override fun support(pushType: String): Boolean = pushType == "TOAST"

    override fun send(targetUserId: String, title: String, content: String) {
        // Logic for Toast popup - e.g., Sending to WebSocket server or specific Kafka topic
        logger.info { "Sending TOAST push to user $targetUserId: $title - $content" }
    }
}

@Component
class ExternalPushSender : PushSender {
    override fun support(pushType: String): Boolean = pushType == "EXTERNAL"

    override fun send(targetUserId: String, title: String, content: String) {
        // Logic for External messaging system (FCM, Slack, etc.)
        logger.info { "Sending EXTERNAL push to user $targetUserId: $title - $content" }
    }
}
